# Machine Failure prediction

Neccessary libraries

In [1]:
import pandas as pd
from sklearn import preprocessing
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split
import keras

_Import the data_

In [2]:
data=pd.read_csv("machine failure.csv")

Take an overveiw of the data

In [3]:
data.head()

UDI Product ID Type  Air temperature [K]  Process temperature [K]  \
0    1     M14860    M                298.1                    308.6   
1    2     L47181    L                298.2                    308.7   
2    3     L47182    L                298.1                    308.5   
3    4     L47183    L                298.2                    308.6   
4    5     L47184    L                298.2                    308.7   

   Rotational speed [rpm]  Torque [Nm]  Tool wear [min]  Machine failure  TWF  \
0                    1551         42.8                0                0    0   
1                    1408         46.3                3                0    0   
2                    1498         49.4                5                0    0   
3                    1433         39.5                7                0    0   
4                    1408         40.0                9                0    0   

   HDF  PWF  OSF  RNF  
0    0    0    0    0  
1    0    0    0    0  
2    0    0    0    0  
3    0    0    0    0  
4    0    0    0    0

Check the Nan (Missing value) from the data

In [4]:
data.isna().sum()

UDI                        0
Product ID                 0
Type                       0
Air temperature [K]        0
Process temperature [K]    0
Rotational speed [rpm]     0
Torque [Nm]                0
Tool wear [min]            0
Machine failure            0
TWF                        0
HDF                        0
PWF                        0
OSF                        0
RNF                        0
dtype: int64

No Missing Values

The type is not the numerical data therefore we need to convert the data in the numeriacl form

In [5]:
data.loc[data["Type"]=='M',"Type"]=1
data.loc[data["Type"]=='L',"Type"]=0
data.loc[data["Type"]=='H',"Type"]=2

data.head()

UDI Product ID Type  Air temperature [K]  Process temperature [K]  \
0    1     M14860    1                298.1                    308.6   
1    2     L47181    0                298.2                    308.7   
2    3     L47182    0                298.1                    308.5   
3    4     L47183    0                298.2                    308.6   
4    5     L47184    0                298.2                    308.7   

   Rotational speed [rpm]  Torque [Nm]  Tool wear [min]  Machine failure  TWF  \
0                    1551         42.8                0                0    0   
1                    1408         46.3                3                0    0   
2                    1498         49.4                5                0    0   
3                    1433         39.5                7                0    0   
4                    1408         40.0                9                0    0   

   HDF  PWF  OSF  RNF  
0    0    0    0    0  
1    0    0    0    0  
2    0    0    0    0  
3    0    0    0    0  
4    0    0    0    0

Devide the data into features and label  (X,y)

In [6]:
data.columns

Index(['UDI', 'Product ID', 'Type', 'Air temperature [K]',
       'Process temperature [K]', 'Rotational speed [rpm]', 'Torque [Nm]',
       'Tool wear [min]', 'Machine failure', 'TWF', 'HDF', 'PWF', 'OSF',
       'RNF'],
      dtype='object')

In [7]:
X=data.drop(columns=['UDI', 'Product ID', 'Machine failure'])
X.head()

Type  Air temperature [K]  Process temperature [K]  Rotational speed [rpm]  \
0    1                298.1                    308.6                    1551   
1    0                298.2                    308.7                    1408   
2    0                298.1                    308.5                    1498   
3    0                298.2                    308.6                    1433   
4    0                298.2                    308.7                    1408   

   Torque [Nm]  Tool wear [min]  TWF  HDF  PWF  OSF  RNF  
0         42.8                0    0    0    0    0    0  
1         46.3                3    0    0    0    0    0  
2         49.4                5    0    0    0    0    0  
3         39.5                7    0    0    0    0    0  
4         40.0                9    0    0    0    0    0

In [8]:
y=data['Machine failure']
y.head()

0    0
1    0
2    0
3    0
4    0
Name: Machine failure, dtype: int64

Check the ranges of the features in X

In [9]:
X.describe()

Air temperature [K]  Process temperature [K]  Rotational speed [rpm]  \
count         10000.000000             10000.000000            10000.000000   
mean            300.004930               310.005560             1538.776100   
std               2.000259                 1.483734              179.284096   
min             295.300000               305.700000             1168.000000   
25%             298.300000               308.800000             1423.000000   
50%             300.100000               310.100000             1503.000000   
75%             301.500000               311.100000             1612.000000   
max             304.500000               313.800000             2886.000000   

        Torque [Nm]  Tool wear [min]           TWF           HDF  \
count  10000.000000     10000.000000  10000.000000  10000.000000   
mean      39.986910       107.951000      0.004600      0.011500   
std        9.968934        63.654147      0.067671      0.106625   
min        3.800000         0.000000      0.000000      0.000000   
25%       33.200000        53.000000      0.000000      0.000000   
50%       40.100000       108.000000      0.000000      0.000000   
75%       46.800000       162.000000      0.000000      0.000000   
max       76.600000       253.000000      1.000000      1.000000   

                PWF           OSF          RNF  
count  10000.000000  10000.000000  10000.00000  
mean       0.009500      0.009800      0.00190  
std        0.097009      0.098514      0.04355  
min        0.000000      0.000000      0.00000  
25%        0.000000      0.000000      0.00000  
50%        0.000000      0.000000      0.00000  
75%        0.000000      0.000000      0.00000  
max        1.000000      1.000000      1.00000

There is a need to Normalize the data 

In [10]:
Scaler=preprocessing.StandardScaler()

In [11]:
X=Scaler.fit_transform(X)

In [12]:
X[:5]

array([[ 7.44412743e-01, -9.52389443e-01, -9.47359888e-01,
         6.81851403e-02,  2.82199756e-01, -1.69598374e+00,
        -6.79798337e-02, -1.07860042e-01, -9.79342412e-02,
        -9.94836193e-02, -4.36304581e-02],
       [-7.45306575e-01, -9.02393410e-01, -8.79959002e-01,
        -7.29471505e-01,  6.33308020e-01, -1.64885170e+00,
        -6.79798337e-02, -1.07860042e-01, -9.79342412e-02,
        -9.94836193e-02, -4.36304581e-02],
       [-7.45306575e-01, -9.52389443e-01, -1.01476077e+00,
        -2.27449840e-01,  9.44289625e-01, -1.61743034e+00,
        -6.79798337e-02, -1.07860042e-01, -9.79342412e-02,
        -9.94836193e-02, -4.36304581e-02],
       [-7.45306575e-01, -9.02393410e-01, -9.47359888e-01,
        -5.90021043e-01, -4.88451785e-02, -1.58600898e+00,
        -6.79798337e-02, -1.07860042e-01, -9.79342412e-02,
        -9.94836193e-02, -4.36304581e-02],
       [-7.45306575e-01, -9.02393410e-01, -8.79959002e-01,
        -7.29471505e-01,  1.31314491e-03, -1.55458761e+00,
  

Divide the data for training, validation and testing

In [13]:
x_train,xt,y_train,yt=train_test_split(X,y,random_state=45,test_size=0.4)
print(x_train.shape,y_train.shape,xt.shape,yt.shape)

(6000, 11) (6000,) (4000, 11) (4000,)


Now divide the data for validation and testing

In [14]:
x_val,x_test,y_val,y_test=train_test_split(xt,yt,random_state=45,test_size=0.5)
print(x_val.shape,y_val.shape,x_test.shape,y_test.shape)

(2000, 11) (2000,) (2000, 11) (2000,)


### Neural network model for the prediction

We will build a model which will have 
+ 7 and 3 perceptron in first and second hidden layer
+ with 1 perceptron in the output layer with the prediction

In [15]:
model=keras.Sequential([
    keras.layers.Input(shape=(11,)),
    keras.layers.Dense(7, activation='relu'),
    keras.layers.Dense(3,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

In [16]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 7)                 84        
                                                                 
 dense_1 (Dense)             (None, 3)                 24        
                                                                 
 dense_2 (Dense)             (None, 1)                 4         
                                                                 
Total params: 112 (448.00 Byte)
Trainable params: 112 (448.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [20]:
model.compile(loss=keras.losses.binary_crossentropy,optimizer=keras.optimizers.Adam(learning_rate=0.01))

In [23]:
model.fit(x_train,y_train,epochs=25)

Epoch 1/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0039
Epoch 2/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0037
Epoch 3/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0033
Epoch 4/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0042
Epoch 5/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0034
Epoch 6/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0034
Epoch 7/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0037
Epoch 8/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0040
Epoch 9/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0039
Epoch 10/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0032
Epoch 11/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0036
Epoch 12/25
188/188 [==============================] - 0s 2ms/step - loss: 0.0030
Epoch 13/25
188/188 [====

In [26]:
x_pred=model.predict(x_val)

63/63 [==============================] - 0s 1ms/step


In [43]:
x_pred[x_pred>=.5]=1
x_pred[x_pred<.5]=0
x_pred=x_pred.reshape(x_pred.shape[0],)
x_pred[:10]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32)

In [54]:
y_val=np.array(y_val)
y_val[:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [56]:
sum(x_pred==y_val)/len(y_val)

0.999

Achieved an accuracy of 99.9% as can be seen from above Check this on test set

In [58]:
x_p=model.predict(x_test)

63/63 [==============================] - 0s 1ms/step


In [65]:
x_p[x_p>=0.5]=1
x_p[x_p<0.5]=0
x_p=x_p.astype("int64").reshape(y_test.shape)
x_p[:10]

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0], dtype=int64)

In [66]:
y_test=np.array(y_test)
y_test[:10]

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0], dtype=int64)

In [67]:
sum(x_p==y_test)/len(y_test)

0.9985

Got 99.85% accuracy on test set